theano/sandbox/gpuarray/opt.py

```diff
@@ -102,28 +97,53 @@ def op_lifter(OP, cuda_only=False):
-
                 # Either one of our inputs is on the gpu or
-                # all of our client are on the gpu
-                if (any([i.owner and i.owner.op == host_from_gpu
-                         for i in node.inputs]) or
-                    all([c != 'output' and c.op == gpu_from_host
-                         for c, idx in node.outputs[0].clients])):
-                    new_op = maker(node)
-                    # This is needed as sometimes new_op inherit from OP.
-                    if new_op and new_op != node.op:
-                        if isinstance(new_op, theano.Op):
-                            return [safe_to_cpu(o) for o in
-                                    new_op(*node.inputs, return_list=True)]
-                        elif isinstance(new_op, (tuple, list)):
-                            return [safe_to_cpu(o) for o in new_op]
-                        else:  # suppose it is a variable on the GPU
-                            return [host_from_gpu(new_op)]
+                # all of our clients are on the gpu
+                replace = False
+                # TODO: Maybe set context_name with infer_context_name()?
+                context_name = None
+                # We replace if any input is a host_from_gpu
+                for i in node.inputs:
+                    if i.owner and i.owner.op == host_from_gpu:
+                        context_name = i.owner.inputs[0].type.context_name
+                        replace = True
+                        break
+                if not replace:
+                    # We replace if *all* clients are on the GPU
+                    clients = [c for o in node.outputs for c in o.clients]
+                    replace = len(clients) != 0
+                    for c, idx in clients:
+                        if (c == 'output' or
+                                not isinstance(c.op, GpuFromHost)):
+                            replace = False
+                    # TODO: check that the clients want the same context?
+                    if replace:
+                        # All clients are GpuFromHost and we have at least one
+                        context_name = clients[0][0].op.context_name
+
+                # Check if we should replace
+                if (not replace or
+                    (cuda_only and
+                     get_context(context_name).kind != 'cuda')):
+                    return False
+
+                new_op = maker(node, context_name)
+                # This is needed as sometimes new_op inherits from OP.
+                if new_op and new_op != node.op:
+                    if isinstance(new_op, theano.Op):
+                        # tag the inputs with the context in case
+                        # the context was derived from the outputs
+                        def tag(i, ctx):
+                            i.tag.context_name = ctx
+                            return i
+                        inputs = [tag(i, context_name) for i in node.inputs]
```

Comment:
The tag can disagree with the type, it's a hint for the op about in which context it should execute.  It doesn't matter if the inputs are not actually in that context.

There may be cases where the inputs are not on the same context and in that case, they will all get transferred to the chosen context.